Sources_v3/StreamChat/Config/Token.swift

```diff
@@ -1,40 +1,61 @@
 //
 
 import Foundation
 
-/// A token is used to authenticate a user.
-public typealias Token = String
+/// The type is designed to store the JWT and the user it is related to.
+public struct Token: Decodable, Equatable, ExpressibleByStringLiteral {
+    public let rawValue: String
+    public let userId: UserId
 
-/// A token provider is a function in which you send a request to your own backend to get a Stream Chat API token.
-///
-/// Set your custom `TokenProvider` in `ChatClientConfig` when creating a `ChatClient` instance. `ChatClient` will use it whenever
-/// it needs to get a new token for the given user.
-///
-public typealias TokenProvider = (_ apiKey: APIKey, _ userId: UserId, _ completion: @escaping (Token?) -> Void) -> Void
+    /// Created a new `Token` instance.
+    /// - Parameter value: The JWT string value. It must be in valid format and contain `user_id` in payload.
+    public init(stringLiteral value: StringLiteralType) {
+        try! self.init(rawValue: value)
+    }
 
-extension Token {
-    /// A token which can be used during development.
-    public static let development: Token = "development"
-    
-    /// Locally checks if the provided token is valid for the given user id.
-    ///
-    /// A token is a string in the JSON Web Token format and one of the parameters it contains is the id of the given user. This
-    /// makes it possible to locally check if the token is valid for the given user.
-    ///
-    /// - Warning: The fact that the token is valid for the given user doesn't mean it can't be rejected by the servers. This is
-    /// a required but not sufficient condition.
-    ///
-    public func isValid(for userId: UserId) -> Bool {
-        if self == .development {
-            return true
+    /// Creates a `Token` instance from the provided `rawValue` if it's valid.
+    /// - Parameter rawValue: The token string in JWT format.
+    /// - Throws: `ClientError.InvalidToken` will be thrown if token string is invalid.
+    public init(rawValue: String) throws {
+        guard let userId = rawValue.jwtPayload?["user_id"] as? String else {
+            throw ClientError.InvalidToken("Provided token does not contain `user_id`")
         }
-        
-        return !userId.isEmpty && (payload?["user_id"] as? String) == userId
+
+        self.init(rawValue: rawValue, userId: userId)
     }
-    
-    var payload: [String: Any]? {
+
+    init(rawValue: String, userId: UserId) {
+        self.rawValue = rawValue
+        self.userId = userId
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+
+        try self.init(
+            rawValue: container.decode(String.self)
+        )
+    }
+}
+
+extension Token {
+    static var anonymous: Self {
+        .init(rawValue: "", userId: .anonymous)
+    }
+
+    static func development(userId: UserId) -> Self {
+        .init(rawValue: "development", userId: userId)
```

Comment:
This was in token provider until I started updating the `Sample` app. There we need a way to create a token to store it in user-defaults without having a token provider that's why how this extension appeared. 

I don't see much harm in it since some users might want to go with:
```swift
.closure { completion in
 sendTokenRequest { result in
   if error {
      completion(.anonymous)
   } else {
      completion(result.token)
   }
 }
}
```

Reply:
@evsaev Makes sense 👍 Lets then add comments explaining when this should be used.

Reply:
added in e55f970b97ce750da86b4692041b7a03c1ee30b7